File: robot_sdk/c/examples/test_robot.c
/**
 * Example test robot using the Robomesh C SDK.
 *
 * Demonstrates: generate keys → provision via admin API → authenticate → heartbeat.
 *
 * Usage:
 *   ./test_robot [host] [tcp_port]
 *   Default: localhost:5001
 */

#include "robomesh.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

static volatile int running = 1;

static void sigint_handler(int sig) {
    (void)sig;
    running = 0;
}

int main(int argc, char *argv[]) {
    const char *host = argc > 1 ? argv[1] : "localhost";
    int tcp_port = argc > 2 ? atoi(argv[2]) : 5001;

    /* 1. Generate keypair */
    robomesh_keypair_t kp;
    if (robomesh_generate_keypair(&kp) != ROBOMESH_OK) {
        fprintf(stderr, "Failed to generate keypair\n");
        return 1;
    }

    char pub_hex[65], priv_hex[65];
    robomesh_public_key_hex(&kp, pub_hex);
    robomesh_private_key_hex(&kp, priv_hex);
    printf("Public key:  %s\n", pub_hex);
    printf("Private key: %s\n", priv_hex);

    /* NOTE: In a real scenario, you'd provision this key via the admin API first.
       For this example, use the pre-seeded example-001 robot instead. */
    printf("\nUsing pre-seeded robot 'example-001' for demo.\n");
    printf("(In production, provision the key above via POST /provision first.)\n\n");

    /* Load the pre-seeded test robot's key */
    robomesh_keypair_t seeded_kp;
    robomesh_load_keypair("c55608b70c4a9f3b43bd1d23e86aaf4c3b2f4b823f54dc34ac668e85363ef2e1",
                           &seeded_kp);

    /* 2. Create client */
    robomesh_config_t config = {
        .uuid = "example-001",
        .host = host,
        .tcp_port = tcp_port,
        .keypair = &seeded_kp,
    };
    robomesh_client_t *client = robomesh_client_create(&config);
    if (!client) {
        fprintf(stderr, "Failed to create client\n");
        return 1;
    }

    /* 3. Authenticate */
    printf("Authenticating...\n");
    robomesh_err_t err = robomesh_authenticate(client);
    if (err != ROBOMESH_OK) {
        fprintf(stderr, "Auth failed: %s\n", robomesh_last_error(client));
        robomesh_client_destroy(client);
        return 1;
    }
    printf("Authenticated! JWT: %.20s...%s\n",
           robomesh_get_jwt(client),
           robomesh_get_jwt(client) + strlen(robomesh_get_jwt(client)) - 10);

    /* 4. Send heartbeats */
    signal(SIGINT, sigint_handler);
    printf("Sending heartbeats. Press Ctrl+C to stop.\n");

    while (running) {
        err = robomesh_send_heartbeat(client, NULL, 60);
        if (err != ROBOMESH_OK) {
            fprintf(stderr, "Heartbeat error: %s\n", robomesh_last_error(client));
            break;
        }
        printf("Heartbeat OK\n");
        sleep(25);
    }

    /* 5. Cleanup */
    printf("Shutting down...\n");
    robomesh_client_destroy(client);
    return 0;
}

File: robot_sdk/python/examples/test_robot.py
#!/usr/bin/env python3
"""Example test robot using the Robomesh SDK.

This script demonstrates the full lifecycle:
1. Generate keys (or use existing ones)
2. Provision via admin API
3. Authenticate via TCP
4. Send heartbeats
5. Exchange messages

Usage:
    # With default dev docker-compose settings:
    python test_robot.py

    # Custom server:
    ROBOMESH_HOST=192.168.1.50 ROBOMESH_TCP_PORT=5002 python test_robot.py
"""

import logging
import os
import signal
import sys
import time

# Add parent to path so we can import without installing
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from robomesh_sdk import RobotClient, generate_ed25519_keypair
from robomesh_sdk.admin import AdminClient

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(name)s] %(message)s")
logger = logging.getLogger("test_robot")

HOST = os.environ.get("ROBOMESH_HOST", "localhost")
TCP_PORT = int(os.environ.get("ROBOMESH_TCP_PORT", "5002"))
HTTP_PORT = int(os.environ.get("ROBOMESH_HTTP_PORT", "8080"))


def main():
    # 1. Generate a fresh keypair
    _, priv_hex, pub_hex = generate_ed25519_keypair()
    robot_uuid = "test-robot-demo"
    device_type = "test_robot"

    logger.info("Robot UUID:    %s", robot_uuid)
    logger.info("Public key:    %s", pub_hex)

    # 2. Provision via admin API
    admin = AdminClient(host=HOST, http_port=HTTP_PORT)
    admin.login("admin", "password1")
    try:
        admin.provision_robot(robot_uuid, pub_hex, device_type)
        logger.info("Robot provisioned successfully")
    except Exception as e:
        logger.warning("Provision failed (may already exist): %s", e)

    # 3. Connect and authenticate
    client = RobotClient(
        uuid=robot_uuid,
        private_key_hex=priv_hex,
        host=HOST,
        tcp_port=TCP_PORT,
    )

    jwt = client.authenticate()
    logger.info("Authenticated! JWT: %s...%s", jwt[:20], jwt[-10:])

    # 4. Register message callback
    def on_message(msg):
        logger.info("Received from server: %s", msg)

    client.on_message(on_message)

    # 5. Send heartbeats in background
    client.start_heartbeat(interval=25, ttl=60)

    # 6. Send some messages
    client.send("hello from test robot")
    time.sleep(1)
    client.send('{"type": "status", "battery": 95, "temp": 37.2}')

    # 7. Wait for Ctrl+C
    logger.info("Robot running. Press Ctrl+C to stop.")

    def handle_sigint(sig, frame):
        logger.info("Shutting down...")
        client.stop_heartbeat()
        client.disconnect()
        sys.exit(0)

    signal.signal(signal.SIGINT, handle_sigint)

    while True:
        time.sleep(1)


if __name__ == "__main__":
    main()
